[Somneel Kumar Saha] iP

SomneelSaha2042 opened the pull request:
# 🚀 TaskMaster: Enhancements & Updates

## **Changes Made**
### 📝 Task Management Enhancements
- ✅ Implemented the `delete` command to remove tasks by index.
- ✅ Added the `help` command to display available commands and their usage.
- 🛠️ Improved error handling for:
  - Invalid commands
  - Out-of-range task indices

### 🧪 Test Cases
- 📄 Created `input.txt` and `EXPECTED.TXT` with test cases for:
  - `todo`
  - `deadline`
  - `event`
  - `list`
- 📊 Verified output correctness by comparing `ACTUAL.TXT` with `EXPECTED.TXT`.

### 📖 README Update
- 🛠️ **Installation & Running Instructions**
- 📜 **List of Available Commands & Descriptions**
- 🗂️ Example usage for `TaskMaster`

---

## **Setup & Usage Instructions**
1. Clone the repository:
   ```sh
   git clone https://github.com/yourusername/taskmaster.git
   cd taskmaster

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SomneelSaha2042 committed:
Level-1

SomneelSaha2042 committed:
Level-1

SomneelSaha2042 committed:
Level 2 Add,List

SomneelSaha2042 committed:
add Level 3

SomneelSaha2042 committed:
add Level 3

SomneelSaha2042 committed:
Level 4

SomneelSaha2042 committed:
Level 5 UI Testing

SomneelSaha2042 committed:
Revert "Level 5 UI Testing"

This reverts commit 49896562e173a61a5a70de300808cedb1a100985.

SomneelSaha2042 committed:
UI testing

SomneelSaha2042 committed:
Level 5 Error Handling

SomneelSaha2042 committed:
Update README

SomneelSaha2042 committed:
Add save functionality

SomneelSaha2042 committed:
Merge branch 'branch-Level-7'

SomneelSaha2042 committed:
Add date and time functionality along with new agenda command

SomneelSaha2042 committed:
improved date time

SomneelSaha2042 committed:
Merge branch 'branch-Level-8'

SomneelSaha2042 committed:
Add JavaDoc comments

SomneelSaha2042 committed:
Made code more Object Oriented and organised into packages

SomneelSaha2042 committed:
Updates README.md and added AI.md

SomneelSaha2042 committed:
Merge branch 'A-MoreOOP'

SomneelSaha2042 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

SomneelSaha2042 committed:
Merge branch 'master' into A-JUnit

SomneelSaha2042 committed:
Add JUnit Tests(have to fix some errors)

SomneelSaha2042 committed:
Merge branch 'A-JUnit'

SomneelSaha2042 committed:
Fixed code(passes JUnit tests) and verified JAR file generation

SomneelSaha2042 committed:
Merge branch 'A-Jar'

SomneelSaha2042 committed:
Added find command

SomneelSaha2042 committed:
Edit README.md

SomneelSaha2042 committed:
Revert "Edit README.md"

This reverts commit da04734442c8f56b4b9f337a512d840465025262.

SomneelSaha2042 committed:
Add CheckStyle config

SomneelSaha2042 committed:
Modify files according CheckStyle(work in progress)

SomneelSaha2042 committed:
Merge branch 'A-CheckStyle'

SomneelSaha2042 committed:
Add JavaFX to TaskMaster and refactor code

SomneelSaha2042 committed:
Add Part 5 of JavaFX tutorial

SomneelSaha2042 committed:
Add images

SomneelSaha2042 committed:
Merge branch 'branch-Level-10'

SomneelSaha2042 committed:
Add multi-keyword find command

SomneelSaha2042 committed:
Merge branch 'A-Varargs'

benedictleejr commented on src/main/java/taskmaster/parser/Parser.java:
```diff
@@ -0,0 +1,163 @@
+                "Invalid date format: " + dateTimeString
+                        + ". Supported formats are: d/M/yyyy HHmm, d-M-yyyy HHmm, ISO_LOCAL_DATE_TIME."
+        );
+    }
```
# Ignoring Exceptions, is this the best way to handle it?
I think having an empty catch block could potentially lead to bugs that are hard to catch down the road 🐛 

An alternative approach could be to first check and verify the format of the input string, and then parse using the appropriate format

SomneelSaha2042 replied:
Hey thanks for catching that! I shall resolve it ASAP

benedictleejr commented on src/main/java/taskmaster/commands/ExitCommand.java:
```diff
@@ -0,0 +1,32 @@
+     * @return {@code true} since this command exits the application.
+     */
+    @Override
+    public boolean isExit() {
```
# Is there a reason this is extracted out into a function?
I feel that there isn't a need to write functions for all the command classes to return a true/false based on whether to exit the program, especially since only the ExitCommand class is supposed to exit.
I'm curious to know if there's a reason for this design 😄 

Alternatively, you could store this as a a boolean attribute, or even handle the logic for exiting in your main app runner class.

SomneelSaha2042 replied:
Its so that isExit can be called on an object of type Command from another class.

benedictleejr reviewed (commented):
LGTM 😉  just some questions and suggestions on code design!

SomneelSaha2042 committed:
Add Lebron

SomneelSaha2042 committed:
Add assertions to enforce assumptions and prevent invalid states

Currently, several methods in the codebase rely on implicit assumptions about input validity.
For example, commands that modify tasks assume that the provided index is within range,
methods in TaskList assume non-null arguments, and Storage assumes the task list is never null when saving.

If these assumptions are violated, the program may behave unpredictably, leading to runtime exceptions
or corrupted state. Without explicit validation, debugging such issues becomes difficult.

Add assertions to critical methods where input validity is assumed.
Ensure that:
- Task indices in `DeleteCommand` are within valid range.
- `AddCommand` only accepts valid task types.
- `Storage.save()` does not receive a null task list.

Assertions are used instead of exceptions because they are intended for debugging and development,
ensuring that assumptions hold during testing. In production, input validation should still be handled
through exception handling.

Assertions are enabled using `-ea` in Java runtime. If disabled, normal execution continues without
performance overhead. The added assertions help catch logic errors early, making debugging easier.

SomneelSaha2042 committed:
Refactor task loading to use Streams for improved efficiency

The task loading mechanism in `Storage` reads task data from a file line by line using a `Scanner` and manually adds each parsed task to a list. This approach works well for small datasets but becomes inefficient when handling large task lists due to iterative processing and the need for explicit list modifications.

As the number of tasks grows, sequential iteration through a `Scanner` can become a bottleneck, leading to performance degradation. Using Streams provides a more declarative and optimized approach for handling large datasets by leveraging parallelism and functional-style operations.

Refactor `Storage.load()` to use Java Streams to read all lines from the file and process each line using `map()` and `collect(Collectors.toList())`. This removes the need for explicit iteration and improves readability and maintainability.

Streams provide a concise and performant way to transform data. The `map()` function allows each line to be processed and converted into a `Task` object, while `collect(Collectors.toList())` efficiently gathers the results into a list.

- Added error handling within the Stream pipeline to skip invalid task entries while logging errors.
- Updated unit tests to verify that task loading functions correctly with the new implementation.
- Verified that performance improves significantly when dealing with large datasets.

SomneelSaha2042 committed:
Add improved help command

SomneelSaha2042 committed:
Fixed Code Quality issues(work in progress)

SomneelSaha2042 committed:
Merge branch 'branch-BCD-Extension'

SomneelSaha2042 committed:
Merge branch 'master' into branch-A-Streams

SomneelSaha2042 committed:
Merge pull request #3 from SomneelSaha2004/branch-A-Streams

Refactor task loading to use Streams for improved efficiency

SomneelSaha2042 committed:
Merge branch 'master' into branch-A-CodeQuality

SomneelSaha2042 committed:
Merge pull request #1 from SomneelSaha2004/branch-A-CodeQuality

Fixed Code Quality issues(work in progress)

SomneelSaha2042 committed:
Merge branch 'master' into branch-A-Assertions

SomneelSaha2042 committed:
Merge pull request #2 from SomneelSaha2004/branch-A-Assertions

Add assertions to enforce assumptions and prevent invalid states

SomneelSaha2042 committed:
Merge branch 'master' of https://github.com/SomneelSaha2004/ip

SomneelSaha2042 committed:
Merge branch 'branch-A-Assertions' of https://github.com/SomneelSaha2004/ip into branch-A-Assertions

SomneelSaha2042 committed:
Merge branch 'branch-A-Assertions'

SomneelSaha2042 committed:
Improve GUI

SomneelSaha2042 committed:
Update README.md

SomneelSaha2042 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	README.md

SomneelSaha2042 committed:
Add Ui.png

SomneelSaha2042 committed:
Fix bugs

SomneelSaha2042 committed:
Fix bug

SomneelSaha2042 committed:
Add Github Pages

SomneelSaha2042 committed:
Update JavaDoc

SomneelSaha2042 committed:
Merge branch 'branch-A-JavaDoc'

SomneelSaha2042 committed:
lebron

SomneelSaha2042 committed:
Add Coding Standard

SomneelSaha2042 committed:
Fix checkstyle

SomneelSaha2042 committed:
Add space

SomneelSaha2042 committed:
Add JavaDoc

SomneelSaha2042 committed:
Edit README.md

SomneelSaha2042 committed:
Fix Bugs

SomneelSaha2042 committed:
Update gradle.build

SomneelSaha2042 committed:
Fix Bugs

SomneelSaha2042 committed:
Fix repository URL in installation instructions
